src/libpcp_web/src/query.c

```diff
@@ -2623,6 +2627,69 @@ series_calculate_rate(node_t *np)
     }
 }
 
+/*
+ * Compare and pick the max instance value(s) among samples.
+ */
+static void
+series_calculate_time_domain_max(node_t *np)
+{
+    seriesQueryBaton	*baton = (seriesQueryBaton *)np->baton;
+    unsigned int	n_series, n_samples, n_instances, i, j, k;
+    double		max_data, data;
+    int			max_pointer;
+    sds			msg;
+
+    n_series = np->left->value_set.num_series;
+    np->value_set.num_series = n_series;
+    np->value_set.series_values = (series_sample_set_t *)calloc(n_series, sizeof(series_sample_set_t));
+    for (i = 0; i < n_series; i++) {
+		n_samples = np->left->value_set.series_values[i].num_samples;
+		if (n_samples > 0){
+			np->value_set.series_values[i].num_samples = n_samples;
+			np->value_set.series_values[i].series_sample = (series_instance_set_t *)calloc(n_samples, sizeof(series_instance_set_t));
+			n_instances = np->left->value_set.series_values[i].series_sample[0].num_instances;
+
+			for (j = 0; j < n_samples; j++){
+				np->value_set.series_values[i].series_sample[j].num_instances = 1;
+				np->value_set.series_values[i].series_sample[j].series_instance = (pmSeriesValue *)calloc(1, sizeof(pmSeriesValue));
+
+				max_pointer = 0;
+				max_data = atof(np->left->value_set.series_values[i].series_sample[j].series_instance[0].data);
+				for (k = 1; k < n_instances; k++){
+					if (np->left->value_set.series_values[i].series_sample[j].num_instances != n_instances) {
+					if (pmDebugOptions.query && pmDebugOptions.desperate) {
+							infofmt(msg, "number of instances in each sample are not equal\n");
+							batoninfo(baton, PMLOG_ERROR, msg);
+						}
+					continue;
+					}                
+					data = atof(np->left->value_set.series_values[i].series_sample[j].series_instance[k].data);
+					if (max_data < data) {
+						max_data = data;
+						max_pointer = k;
+					}
+				}
+				
+				np->value_set.series_values[i].series_sample[j].series_instance[0].timestamp = 
+					sdsnew(np->left->value_set.series_values[i].series_sample[j].series_instance[max_pointer].timestamp);
+				np->value_set.series_values[i].series_sample[j].series_instance[0].series = 
+					sdsnew(np->left->value_set.series_values[i].series_sample[j].series_instance[max_pointer].series);
+				np->value_set.series_values[i].series_sample[j].series_instance[0].data = 
+					sdsnew(np->left->value_set.series_values[i].series_sample[j].series_instance[max_pointer].data);
+				np->value_set.series_values[i].series_sample[j].series_instance[0].ts = 
+					np->left->value_set.series_values[i].series_sample[j].series_instance[max_pointer].ts;
```

Comment:
There's a code pattern I've noticed developing throughout query.c which I'm not sure is ideal, where we have these very long statements repeatedly de-referencing pointers (I know you're just continuing the same pattern here & usually that's the best approach to take).  I wonder if it would make the code more readable, however, if we used some intermediate local variables.

For example here we repeat the: "np->value_set.series_values[i].series_sample[j].series_instance[0]" string several times, as well as "np->left->value_set.series_values[i].series_sample[j].series_instance[max_pointer]".  Shall we start moving code over to assigning earlier on to a local variable (e.g. "inst", and then using "inst->timestamp = ...; inst->series = ...; " and so on) ... ?

Reply:
Thank you for your suggestions, and I agree with you that defining some local variables like "inst" will improve the readability of our code. I will update this.